src/map.cpp

```diff
@@ -1775,6 +1822,253 @@ int map::climb_difficulty( const tripoint &p ) const
     return best_difficulty - blocks_movement;
 }
 
+bool map::has_floor( const tripoint &p ) const
+{
+    return !zlevels ||
+           p.z < -OVERMAP_DEPTH + 1 || p.z > OVERMAP_HEIGHT ||
+           !has_flag( TFLAG_NO_FLOOR, p );
+}
+
+bool map::supports_above( const tripoint &p ) const
+{
+    const maptile tile = maptile_at( p );
+    const ter_t &ter = tile.get_ter_t();
+    if( ter.movecost == 0 ) {
+        return true;
+    }
+
+    const furn_id frn_id = tile.get_furn();
+    if( frn_id != f_null ) {
+        const furn_t &frn = frn_id.obj();
+        if( frn.movecost < 0 ) {
+            return true;
+        }
+    }
+
+    if( veh_at( p ) != nullptr ) {
+        return true;
+    }
+
+    return false;
+}
+
+void map::drop_everything( const tripoint &p )
+{
+    if( has_floor( p ) ) {
+        return;
+    }
+
+    drop_furniture( p );
+    drop_items( p );
+    drop_vehicle( p );
+}
+
+void map::drop_furniture( const tripoint &p )
+{
+    const furn_id frn = furn( p );
+    if( frn == f_null ) {
+        return;
+    }
+
+    enum support_state {
+        SS_NO_SUPPORT = 0,
+        SS_BAD_SUPPORT, // TODO: Implement bad, shaky support
+        SS_GOOD_SUPPORT,
+        SS_FLOOR, // Like good support, but bash floor instead of tile below
+        SS_CREATURE
+    };
+
+    // Checks if the tile:
+    // has floor (supports unconditionally)
+    // has support below
+    // has unsupporting furniture below (bad support, things should "slide" if possible)
+    // has no support and thus allows things to fall through
+    const auto check_tile = [this]( tripoint &pt ) {
+        tripoint dest( pt );
+        if( has_floor( dest ) ) {
+            pt = dest;
+            return SS_FLOOR;
+        }
+
+        tripoint below_dest( pt.x, pt.y, pt.z - 1 );
+        if( supports_above( below_dest ) ) {
+            pt = dest;
```

Comment:
That assignment does not seem to do anything (same for the one above and at the end of the function).

`dests` is initialized to be a copy of `pt` and neither of those points is changed here. it seems like a no-op and should be removed.